File: openclaw_py/gateway/http_common.py
"""HTTP response helper functions.

This module provides utilities for sending standardized HTTP responses.
"""

from typing import Any

from fastapi import Response
from fastapi.responses import JSONResponse


def send_json(status_code: int, body: Any) -> JSONResponse:
    """Send a JSON response.

    Args:
        status_code: HTTP status code
        body: Response body (will be serialized to JSON)

    Returns:
        JSONResponse object

    Examples:
        >>> response = send_json(200, {"status": "ok"})
        >>> response.status_code
        200
    """
    return JSONResponse(content=body, status_code=status_code)


def send_text(body: str, status_code: int = 200) -> Response:
    """Send a plain text response.

    Args:
        body: Response text
        status_code: HTTP status code (default: 200)

    Returns:
        Response object
    """
    return Response(content=body, media_type="text/plain", status_code=status_code)


def send_unauthorized(message: str = "Unauthorized") -> JSONResponse:
    """Send a 401 Unauthorized response.

    Args:
        message: Error message

    Returns:
        JSONResponse with 401 status
    """
    return JSONResponse(
        content={
            "error": {
                "message": message,
                "type": "unauthorized",
            }
        },
        status_code=401,
    )


def send_invalid_request(message: str) -> JSONResponse:
    """Send a 400 Bad Request response.

    Args:
        message: Error message

    Returns:
        JSONResponse with 400 status
    """
    return JSONResponse(
        content={
            "error": {
                "message": message,
                "type": "invalid_request_error",
            }
        },
        status_code=400,
    )


def send_not_found(message: str = "Not Found") -> JSONResponse:
    """Send a 404 Not Found response.

    Args:
        message: Error message

    Returns:
        JSONResponse with 404 status
    """
    return JSONResponse(
        content={
            "error": {
                "message": message,
                "type": "not_found_error",
            }
        },
        status_code=404,
    )


def send_method_not_allowed(method: str = "POST") -> JSONResponse:
    """Send a 405 Method Not Allowed response.

    Args:
        method: The method that was attempted

    Returns:
        JSONResponse with 405 status
    """
    return JSONResponse(
        content={
            "error": {
                "message": f"Method {method} not allowed",
                "type": "method_not_allowed",
            }
        },
        status_code=405,
    )

File: src/gateway/http-common.ts
import type { IncomingMessage, ServerResponse } from "node:http";

import { readJsonBody } from "./hooks.js";

// -- Security headers --

/** Apply baseline security headers to every HTTP response. */
export function setSecurityHeaders(res: ServerResponse) {
  res.setHeader("X-Content-Type-Options", "nosniff");
  res.setHeader("X-Frame-Options", "DENY");
  // Disable legacy XSS auditor (can introduce side-channel attacks)
  res.setHeader("X-XSS-Protection", "0");
  res.setHeader("Referrer-Policy", "strict-origin-when-cross-origin");
  res.setHeader("Permissions-Policy", "camera=(), microphone=(), geolocation=()");
  res.setHeader("Cache-Control", "no-store");
}

/** Add HSTS header for TLS connections. */
export function setHstsHeader(res: ServerResponse) {
  res.setHeader("Strict-Transport-Security", "max-age=31536000; includeSubDomains");
}

// -- CORS --

export type CorsConfig = {
  allowedOrigins?: string[];
  allowedMethods?: string[];
  maxAge?: number;
};

/**
 * Handle CORS for a request.
 * Returns "preflight" if the request was an OPTIONS preflight (response already sent),
 * or "continue" if the handler chain should proceed.
 * When no allowedOrigins are configured, no CORS headers are set (deny cross-origin by default).
 */
export function handleCors(
  req: IncomingMessage,
  res: ServerResponse,
  config?: CorsConfig,
): "preflight" | "continue" {
  const origin = req.headers.origin;
  if (!origin || !config?.allowedOrigins || config.allowedOrigins.length === 0) {
    // No origin header or no allowed origins configured — deny cross-origin silently
    if (req.method === "OPTIONS") {
      res.statusCode = 204;
      res.end();
      return "preflight";
    }
    return "continue";
  }

  const allowed = config.allowedOrigins.includes("*") || config.allowedOrigins.includes(origin);
  if (!allowed) {
    if (req.method === "OPTIONS") {
      res.statusCode = 204;
      res.end();
      return "preflight";
    }
    return "continue";
  }

  res.setHeader("Access-Control-Allow-Origin", origin);
  res.setHeader("Vary", "Origin");

  if (req.method === "OPTIONS") {
    const methods = config.allowedMethods ?? ["GET", "POST", "OPTIONS"];
    res.setHeader("Access-Control-Allow-Methods", methods.join(", "));
    res.setHeader("Access-Control-Allow-Headers", "Content-Type, Authorization, X-Moltbot-Token");
    res.setHeader("Access-Control-Max-Age", String(config.maxAge ?? 86400));
    res.statusCode = 204;
    res.end();
    return "preflight";
  }

  return "continue";
}

export function sendJson(res: ServerResponse, status: number, body: unknown) {
  res.statusCode = status;
  res.setHeader("Content-Type", "application/json; charset=utf-8");
  res.end(JSON.stringify(body));
}

export function sendText(res: ServerResponse, status: number, body: string) {
  res.statusCode = status;
  res.setHeader("Content-Type", "text/plain; charset=utf-8");
  res.end(body);
}

export function sendMethodNotAllowed(res: ServerResponse, allow = "POST") {
  res.setHeader("Allow", allow);
  sendText(res, 405, "Method Not Allowed");
}

export function sendUnauthorized(res: ServerResponse) {
  sendJson(res, 401, {
    error: { message: "Unauthorized", type: "unauthorized" },
  });
}

export function sendInvalidRequest(res: ServerResponse, message: string) {
  sendJson(res, 400, {
    error: { message, type: "invalid_request_error" },
  });
}

export async function readJsonBodyOrError(
  req: IncomingMessage,
  res: ServerResponse,
  maxBytes: number,
): Promise<unknown> {
  const body = await readJsonBody(req, maxBytes);
  if (!body.ok) {
    sendInvalidRequest(res, body.error);
    return undefined;
  }
  return body.value;
}

export function writeDone(res: ServerResponse) {
  res.write("data: [DONE]\n\n");
}

export function setSseHeaders(res: ServerResponse) {
  res.statusCode = 200;
  res.setHeader("Content-Type", "text/event-stream; charset=utf-8");
  res.setHeader("Cache-Control", "no-cache");
  res.setHeader("Connection", "keep-alive");
  res.flushHeaders?.();
}
